Refactor documents/mango addon to use Redux

Antonio-Maranhao opened the pull request:
## Overview

Updated mango query to use Redux

Pending:
 - [x] Replace all uses of `typeOfIndex` with `docType`
 - [x] Introduce new `docType` for Mango query results
 - [x] Hide 'Metadata' button for Mango queries
 - [x] Remove `dispatch` from api.js files so they are not aware of Redux of any frameworks
 - [x] Cleanup unused code
 - [x] Resolve conflicts from master

## Testing recommendations

- Run Mango query
   - Switch between Table / JSON modes
   - Select & delete docs
   - Change 'Documents per page'
   - Navigate between pages using '<' and '>' buttons
- Click on 'manage index'
   - Index list should be displayed in the JSON mode
   - Create a new index. Check that index list is updated
   - Select & delete index
   - Change 'Documents per page'
   - Navigate between pages using '<' and '>' buttons

## Checklist

- [x] Code is written and works correctly;
- [ ] Changes are covered by tests;
- [x] Documentation reflects the changes;

Antonio-Maranhao commented:
@garrensmith This branch works (all tests pass) but I still need to merge Will's changes for query history and index templates. 
Not sure if he has anything more coming down the pipe - if so it'd be worth to hold on until we can merge this PR, otherwise I'll keep rewriting his new code.
Thanks

garrensmith commented on app/addons/documents/header/header.js:
```diff
@@ -31,7 +31,8 @@ export default class BulkDocumentHeaderController extends React.Component {
       bulkDocCollection: indexResultsStore.getBulkDocCollection(),
       isMango: indexResultsStore.getIsMangoResults(),
-      isMangoIndexList: indexResultsStore.getIsMangoIndexResults()
+      //isMangoIndexList: indexResultsStore.getIsMangoIndexResults()
```
I guess you can remove

garrensmith commented on app/addons/documents/header/header.js:
```diff
@@ -62,14 +63,14 @@ export default class BulkDocumentHeaderController extends React.Component {
     const layout = changeLayout ? selectedLayout : this.state.selectedLayout;
     let metadata, json, table;
-    if ((typeOfIndex && typeOfIndex === 'view') || !this.state.isMango) {
+    if ((typeOfIndex && typeOfIndex === 'view')) {
```
Couldn't this be simplified to 
`if (typeOfIndex === 'view')`

garrensmith commented on app/addons/documents/header/header.js:
```diff
@@ -62,14 +63,14 @@ export default class BulkDocumentHeaderController extends React.Component {
           Metadata
         </Button>;
-    } else if (this.state.isMangoIndexList) {
+    } else if ((typeOfIndex && typeOfIndex === 'MangoIndex')) {
```
This can be implified to `if (typeOfIndex === 'MangoIndex')`

garrensmith commented on app/addons/documents/index-results/apis/base.js:
```diff
@@ -24,12 +25,13 @@ export const resetState = () => {
 };
 
-export const newResultsAvailable = (docs, params, canShowNext) => {
+export const newResultsAvailable = (docs, params, canShowNext, docType) => {
```
I've been wondering about this. I don't think we should have any `action` functions in the api. The `api` files should just do ajax queries and be a basic api to interact with CouchDB. I don't think the `api's` should know about redux or anything like that. They should just pass data and then the action functions in the `action.js` files should do any required dispatching. What do you think?

garrensmith commented on app/addons/documents/index-results/apis/fetch.js:
```diff
@@ -112,7 +116,12 @@ export const queryEndpoint = (fetchUrl, params) => {
-  .then(res => res.error ? [] : res.rows);
+  .then(res => {
+    return {
+      docs: res.error ? [] : res.rows,
```
Do we show the error if there is an error?

garrensmith commented on app/addons/documents/index-results/apis/fetch.js:
```diff
@@ -183,14 +198,30 @@ export const postToBulkDocs = (databaseName, payload) => {
-  const failedDocs = res.filter(doc => !!doc.error).map(doc => doc.id);
-  const hasDesignDocs = !!originalDocs.map(d => d._id).find((_id) => /_design/.test(_id));
+  const failedDocs = docType === 'MangoIndex' ?
+    (res.fail ? res.fail.map(doc => doc.id) : [])
```
I think lets expand this. Its pretty tricky to read.

garrensmith commented on app/addons/documents/index-results/components/results/IndexResults.js:
```diff
@@ -20,25 +20,29 @@ export default class IndexResults extends React.Component {
   }
 
   componentWillUpdate (nextProps) {
+    // console.log('IndexResults - will update - next.ddocsOnly:', nextProps.ddocsOnly);
```
You can remove

garrensmith commented on app/addons/documents/index-results/helpers/shared-helpers.js:
```diff
@@ -76,10 +76,33 @@ const getDefaultPerPage = () => {
+  return !doc || !doc.attributes || !Object.keys(doc.attributes).length;
+};
+
+const getDocId = (doc, docType = 'view') => {
```
Nice idea here

garrensmith commented on app/addons/documents/mango/mango.api.js:
```diff
@@ -0,0 +1,114 @@
+      'Accept': 'application/json',
+      'Content-Type': 'application/json; charset=utf-8'
+    },
+    credentials: 'include',
```
This is the perfect api.js file. I think we need to make them all like this

Antonio-Maranhao commented:
@garrensmith All done. Could you please review?

garrensmith commented on app/addons/documents/index-results/reducers.js:
```diff
@@ -211,11 +222,14 @@ export const getAllDocsSelected = (state) => {
 
   for (let i = 0; i < state.docs.length; i++) {
     const doc = state.docs[i];
```
This for loop seems really complicated. Could we simplify it move by using a `.reduce` or `.filter`?

garrensmith commented on app/addons/documents/mango/components/MangoIndexEditorContainer.js:
```diff
@@ -0,0 +1,48 @@
+  return {
+    saveIndex: (options) => {
+      dispatch(Actions.saveIndex(options));
+      // dispatch(IndexResultActions.fetchDocs(queryDocs, options.fetchParams, {}));
```
Can we remove this?

garrensmith commented on app/addons/documents/mango/components/MangoQueryEditorContainer.js:
```diff
@@ -0,0 +1,99 @@
+  if (!availableIndexes) {
+    return [];
+  }
+  return availableIndexes.filter(function ({ type }) {
```
Can you use `() =>` and also `array.includes

garrensmith commented on app/addons/documents/mango/components/MangoQueryEditorContainer.js:
```diff
@@ -0,0 +1,99 @@
+    return;
+  }
+
+  return additionalIndexes.filter(function (el) {
```
fat arrow

garrensmith commented on app/addons/documents/mango/mango.reducers.js:
```diff
@@ -0,0 +1,164 @@
+};
+
+const updateQueryHistory = ({ historyKey, history }, queryCode, label) => {
+  //const history = getHistory();
```
I think we can remove

garrensmith commented on app/addons/documents/index-results/reducers.js:
```diff
@@ -85,17 +86,27 @@ export default function resultsState (state = initialState, action) {
       });
 
     case ActionTypes.INDEX_RESULTS_REDUX_CHANGE_LAYOUT:
+      console.log('Reducing to ', action.layout);
```
Can you remove

garrensmith commented:
I've noticed two issues while test:
* If I edit the selector for a query and then go to a different database, the selector query is still the changed selector instead of changing back to the default `_id` one.
* If I run a query, paginate to the next page. Then change the selector and query again. It keeps me on the second page instead of reseting to page one for a new query

garrensmith commented:
Just the two issues. Other than that, this is looking really good. Great work.

Antonio-Maranhao commented:
@garrensmith Addressed your comments and fixed both issues. For the 1st issue, it'll load the last query from the exec history for that database, otherwise it loads the default one with '_id'.

garrensmith commented:
@Antonio-Maranhao nice work. I found another bug. If I set the limit it doesn't obey that limit.
So if set a `limit: 5` and set it to a page limit of 5 docs per page. It will still do a query with a limit of 6 and allow me to paginate to another page.
